File: src/util/deps.ts
import { run, which, whichAny } from "./exec.js";
import { log } from "./logger.js";

// Detect available package runtimes and a usable npx-equivalent.

export interface RuntimeReport {
  node: string | null;
  npm: string | null;
  npx: string | null;
  bun: string | null;
  bunx: string | null;
  cargo: string | null;
  python: string | null;
}

export function detectRuntimes(): RuntimeReport {
  return {
    node: which("node"),
    npm: which("npm"),
    npx: which("npx"),
    bun: which("bun"),
    bunx: which("bunx"),
    cargo: which("cargo"),
    python: which("python3") ?? which("python"),
  };
}

export function pickJsRunner(): { cmd: string; runArgs: string[] } | null {
  const npx = which("npx");
  if (npx) return { cmd: "npx", runArgs: ["-y"] };
  const bunx = which("bunx");
  if (bunx) return { cmd: "bunx", runArgs: [] };
  return null;
}

export async function ensureNpmInstall(pkg: string, opts: { global?: boolean } = {}): Promise<boolean> {
  const npm = which("npm");
  const bun = which("bun");
  if (npm) {
    const args = ["install", opts.global ? "-g" : "", `${pkg}@latest`].filter(Boolean);
    const r = await run("npm", args);
    return r.code === 0;
  }
  if (bun) {
    const args = ["add", opts.global ? "-g" : "", `${pkg}@latest`].filter(Boolean);
    const r = await run("bun", args);
    return r.code === 0;
  }
  log.err(`Neither npm nor bun is installed — cannot install ${pkg}.`);
  return false;
}

export function reportRuntimes(r: RuntimeReport): void {
  const fmt = (name: string, p: string | null) =>
    p ? log.sub(`${name}: ${p}`) : log.sub(`${name}: not found`);
  fmt("node", r.node);
  fmt("npm/npx", r.npm ?? r.npx);
  fmt("bun/bunx", r.bun ?? r.bunx);
  fmt("cargo", r.cargo);
  fmt("python", r.python);
}

File: internal/util/deps.go
package util

// RuntimeReport lists discovered package runtimes (path or "").
type RuntimeReport struct {
	Node, Npm, Npx, Bun, Bunx, Cargo, Python string
}

func DetectRuntimes() RuntimeReport {
	py := Which("python3")
	if py == "" {
		py = Which("python")
	}
	return RuntimeReport{
		Node:   Which("node"),
		Npm:    Which("npm"),
		Npx:    Which("npx"),
		Bun:    Which("bun"),
		Bunx:   Which("bunx"),
		Cargo:  Which("cargo"),
		Python: py,
	}
}

// PickJsRunner prefers npx, then bunx.
func PickJsRunner() (cmd string, runArgs []string, ok bool) {
	if Which("npx") != "" {
		return "npx", []string{"-y"}, true
	}
	if Which("bunx") != "" {
		return "bunx", []string{}, true
	}
	return "", nil, false
}

func ReportRuntimes(r RuntimeReport) {
	fmtLine := func(name, p string) {
		if p != "" {
			L.Sub(name + ": " + p)
		} else {
			L.Sub(name + ": not found")
		}
	}
	fmtLine("node", r.Node)
	npmnpx := r.Npm
	if npmnpx == "" {
		npmnpx = r.Npx
	}
	fmtLine("npm/npx", npmnpx)
	bunbunx := r.Bun
	if bunbunx == "" {
		bunbunx = r.Bunx
	}
	fmtLine("bun/bunx", bunbunx)
	fmtLine("cargo", r.Cargo)
	fmtLine("python", r.Python)
}
